src/SubtitlesParser/index.js

```diff
@@ -0,0 +1,189 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Log } from '../LightningSdkPlugins'
+
+export default class SubtitlesParser {
+  // @ params url: subtitle file URL
+  // @ customParser: a customParser to use instead of default parser of the plugin
+  // @ parseOptions.removeSubtitleTextStyles: remove subtitle textstyles possible value true or false
+  // @return parsed subtitles as list of objects
+  // also stores parsed data
+  static fetchAndParseSubs(
+    url,
+    customParser = false,
+    parseOptions = { removeSubtitleTextStyles: true }
+  ) {
+    const _url = new URL(url)
+    if (!((_url.protocol === 'https:' || _url.protocol === 'http:') && _url.hostname)) {
+      Log.info('Invalid subtitle Url')
+      return Promise.reject(new Error('Invalid URL'))
+    }
+
+    return new Promise((resolve, reject) => {
+      fetch(url)
+        .then(data => {
+          let subtitleData = data.text()
+
+          this.clearAllSubtitles()
+          if (customParser && typeof customParser === 'function') {
+            this._captions = customParser(subtitleData)
+          } else {
+            this._captions = this.parseSubtitles(subtitleData, parseOptions)
+          }
+          if (this._captions && this._captions.length) {
+            resolve(this._captions)
+          } else {
+            reject('Failed to parse subtitles: invalid subtitles length')
+          }
+        })
+        .catch(error => {
+          Log.error('Fetching subtitles file Failed:', error)
+          this.clearAllSubtitles()
+          reject('Fetching subtitles file Failed')
+        })
+    })
+  }
+
+  // clears stored subtitles data
+  static clearAllSubtitles() {
+    this._captions = null
+  }
+
+  // get current subtitles
+  // @ currentTime: currentTime in seconds
+  // @return: subtitle text at that currentTime
+  static getSubtitleByTimeIndex(currentTime) {
+    if (currentTime === undefined || isNaN(currentTime) || typeof currentTime !== 'number') {
+      throw new Error('You should pass a currentTime to fetch the current subtitle')
+    }
+
+    if (!Array.isArray(this._captions) || this._captions.length <= 0) {
+      throw new Error("didn't find and stored captions in plugin")
+    }
+
+    if (this._lastIndex > this._captions.length - 1 || !this._lastIndex) {
+      this._lastIndex = 0
+    }
+    const activeIndex = this.getActiveIndex(currentTime) // find active cue from the captions stored
+
+    if (activeIndex !== -1 && activeIndex <= this._captions.length - 1) {
+      return this._captions[activeIndex].payload
+    } else if (activeIndex === -1) {
+      return ''
+    }
+  }
+
+  static getActiveIndex(currentTime) {
```

Comment:
As a performance optimisation, you could `Math.floor` the currentTime and keep track of the last time (in seconds, or intervals of 0.5 seconds) that was used to get the cue for. Sort of like memoization. Because this function will be fired on veery `timeupdate` event (which occurs a lot), I would try to deduplicate the amount of times it actually executes this function a bit

Reply:
Calling getActiveIndex for every 0.5-second change in currentTime from `timeupdate` event.